libs/ml/include/ml/optimization/optimizer.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/graph.hpp"
+#include "ml/ops/loss_functions/criterion.hpp"
+
+namespace fetch {
+namespace ml {
+
+/**
+ * Abstract gradient optimizer class
+ * @tparam T ArrayType
+ * @tparam C CriterionType
+ */
+template <class T, class C>
+class Optimizer
+{
+public:
+  using ArrayType     = T;
+  using CriterionType = C;
+  using DataType      = typename ArrayType::Type;
+  using SizeType      = typename ArrayType::SizeType;
+
+  Optimizer(std::shared_ptr<Graph<T>> graph, std::string const &input_node_name,
+            std::string const &output_node_name, DataType const &learning_rate)
+    : graph_(graph)
+    , input_node_name_(input_node_name)
+    , output_node_name_(output_node_name)
+    , learning_rate_(learning_rate)
+  {}
+
+  DataType DoBatch(ArrayType &data, ArrayType &labels)
+  {
+    DataType loss{0};
+    SizeType n_data = data.shape().at(0);
+
+    for (SizeType step{0}; step < n_data; ++step)
+    {
+      auto cur_input = data.Slice(step).Copy();
```

Comment:
this slice.copy is expensive. I recommend that for now you leave this implementation but please raise an issue that this needs to be optimised

Reply:
Slice should give you something that will behave exactly like Tensor rather than TensorSlice.